### ISSUE TITLE
WebSocketClientFactory's _batched_timer SHOULD use the passed event loop

### ISSUE THREAD

**Issue by karolyi** [crossbario/autobahn-python#1007]
Hey,

here's my second issue report in a week:

The `WebSocketClientFactory` can be passed an event loop:

https://github.com/crossbario/autobahn-python/blob/bba4ed8418e59a678012ecb466092bee250b906f/autobahn/asyncio/websocket.py#L288-L289

.... and `WebSocketClientFactory`'s `_batched_timer` **SHOULD** use it, which it doesn't now:
https://github.com/crossbario/autobahn-python/blob/bba4ed8418e59a678012ecb466092bee250b906f/autobahn/websocket/protocol.py#L3798-L3801

Why is this a problem? Because if someone doesn't have a main event loop, autoPings won't work. My application doesn't have a **main** event loop, it only has an event loop in a separate thread, which I initialize later, along with the websocket client.

As far as I've seen, `txaio` uses the main event loop, and the websocket client can't be told to use other than the main initialized one.

Correct me if I'm wrong, but now I have to have an asynchronous main event loop, just to have autopings working?

**Comment by meejah** [comment]
Some overlap with https://github.com/crossbario/autobahn-python/issues/747

For context, this is actually a fairly deep issue since `txaio` is based around a "lowest common denominator" papering-over of the differences between Twisted and asyncio. Twisted has a single global reactor, and so does txaio.

In this particular case, I think passing `loop=` onwards to `make_batched_timer` is definitely "more correct", but also this needs a more comprehensive story, I think. Another option is to configure txaio to use your loop instead of passing a `loop=` kwarg to the factory -- that is: `txaio.config.loop = loop`?

(You're right that "txaio" has ONE event-loop, but it can be any eventloop; the default is `asyncio.get_event_loop()`...)

**Comment by karolyi** [comment]
I agree with the deep statement. I took me a couple hours to get down to the bottom of this. I have it now in 'worksforme' state, but it can be confusing for newcomers, and I think it should be improved in the long run. Maybe my original issue comment will help those.

**Comment by meejah** [comment]
Hmm, I was going to make the one-line fix but this is the base-class and the Twisted one doesn't have a `self.loop`..so not quite the oner-liner I was hoping :/

The "real" fix here is to decide on ONE way to configure event-loops. The easiest would be to proclaim that "`txaio` has a single loop, and you can set it to whatever you want" (as now) and deprecate usage of any `loop=` kwargs. I'm mostly a "Twisted person" but it seems like many asyncio libraries don't really deal properly with > 1 event-loop via `loop=` kwargs anyway.

If we do want to allow the use of `loop=` kwargs, then they need to correctly deal with multiple event-loops (and I don't really understand the use-cases for having multiple event-loops). So, writing those down somewhere would be good..
@oberstet

**Comment by meejah** [comment]
(BTW, thanks for tracking this down and for a clear bug report!)

**Comment by karolyi** [comment]
> Hmm, I was going to make the one-line fix but this is the base-class and the Twisted one doesn't have a self.loop..so not quite the oner-liner I was hoping :/

Indeed. I was looking at the twisted version of this function too, and then saw that adding the event loop to it is not a one single-and-easy patch. Maybe using a default argument on `make_batched_timer` so that it would default to its own loop, can be useful.

In my opinion, having separate event loops passed in the case of asyncio shouldn't be a problem, as long as they're running.

Having multiple event loops is just a question of taste. Since my application doesn't have a 'main' event loop (as it's synchronous, a Kivy app), I have to create my own for the thread that communicates with my server using the websocket client.

**Comment by meejah** [comment]
I get that "custom" loops have use-cases, I just don't understand why you'd want 2 or more event-loops.

**Comment by mwcraig** [comment]
One minor comment for anyone else who runs into it: I had to also `async.set_event_loop(my_loop)` in addition to `txaio.config.loop = my_loop`

My use case is that software out of my control had already started an event loop so I needed to make my own in a separate thread.
